certora/specs/AllocationMorphoVaultV1Adapter.spec

```
@@ -0,0 +1,153 @@
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+
+    function _.market(MorphoHarness.Id) external => DISPATCHER(true);
+    function _.idToMarketParams(MorphoHarness.Id) external => DISPATCHER(true);
+    function _.position(MorphoHarness.Id, address) external => DISPATCHER(true);
+    function _.supplyShares(MorphoHarness.Id, address) external => DISPATCHER(true);
+    function _.accrueInterest(MorphoHarness.MarketParams) external => DISPATCHER(true);
+    function _.supply(MorphoHarness.MarketParams, uint256, uint256, address, bytes) external => DISPATCHER(true);
+    function _.withdraw(MorphoHarness.MarketParams, uint256, uint256, address, address) external => DISPATCHER(true);
+}
+
+function mulDivSummary(uint256 x, uint256 y, uint256 denominator) returns uint256 {
+    mathint result;
+    if (denominator == 0) revert();
+    result = x * y / denominator;
+    if (result >= 2^256) revert();
+    return assert_uint256(result);
+}
+
+persistent ghost uint256 constantBorrowRate;
+
+persistent ghost int256 ghostChange;
+
+// Wrapper to record interest value returned by the adapter.
+function morphoVaultV1AdapterWrapperSummary(env e, bool isAllocateCall, bytes data, uint256 assets, bytes4 bs, address a) returns (bytes32[], int256) {
+    bytes32[] ids;
+    int256 change;
+
+    if (isAllocateCall) {
+        (ids, change) = MorphoVaultV1Adapter.allocate(e, data, assets, bs, a);
+    } else {
+        (ids, change) = MorphoVaultV1Adapter.deallocate(e, data, assets, bs, a);
+    }
+
+    ghostChange = change;
+
+    return (ids, change);
+}
+
+rule allocateMorphoVaultV1Adapter(env e, bytes data, uint256 assets) {
+    // Trick to require that all the following addresses are different.
+    require (MorphoVaultV1 == 0x10, "ack");
+    require (MorphoVaultV1Adapter == 0x11, "ack");
+    require (currentContract == 0x12, "ack");
+
+
+    // Ensure the VaultV2 and MorphoVaultV1 contracts are properly linked to the adapter in the conf file.
+    assert MorphoVaultV1Adapter.parentVault == currentContract;
+    assert MorphoVaultV1Adapter.morphoVaultV1 == MorphoVaultV1;
+
+    bytes32[] adapterIds = MorphoVaultV1Adapter.ids();
+
+    uint i;
+    require (i < adapterIds.length, "require i to be a valid index");
+
+    bytes32 id;
+    uint256 allocationBefore = allocation(id);
+
+    requireInvariant allocationIsInt256(adapterIds[i]);
+    int256 idIAllocationBefore = assert_int256(allocation(adapterIds[i]));
+
+    allocate(e, MorphoVaultV1Adapter, data, assets);
+
+    assert allocation(adapterIds[i]) == idIAllocationBefore + ghostChange;
+    assert !(exists uint j . j < adapterIds.length && id == adapterIds[j]) => currentContract.caps[id].allocation == allocationBefore;
```

```suggestion
    assert currentContract.caps[id].allocation != allocationBefore => exists uint j. j < adapterIds.length && id == adapterIds[j];
```
[optional] I find it a bit clearer, "only changes are in the id list"